extensions/wikia/ContributionAppreciation/ContributionAppreciationController.class.php

```diff
@@ -3,13 +3,46 @@
 class ContributionAppreciationController extends WikiaController {
 
 	public function appreciate() {
-		global $wgUser;
+		$user = $this->wg->User;
 
-		if ( $this->request->isValidWriteRequest( $wgUser ) ) {
-			//TODO: do something with apprecation
+		$this->request->isValidWriteRequest( $user );
+
+		$revisionId = $this->request->getInt( 'revision' );
+		$revision = Revision::newFromId( $revisionId );
+		if ( $revision instanceof Revision ) {
+			( new RevisionUpvotesService() )->addUpvote(
+				$this->wg->CityId,
+				$revision->getPage(),
+				$revisionId,
+				$revision->getUser(),
+				$user->getId()
+			);
 		}
 	}
 
+	public function getAppreciations() {
+		$html = '';
+		$user = $this->wg->User;
+		$upvotesService = new RevisionUpvotesService();
+		$upvotes = $upvotesService->getUserNewUpvotes( $user->getId() );
+
+		if ( !empty( $upvotes ) ) {
+			$appreciations = $this->prepareAppreciations( $upvotes );
```

Comment:
Is it possible to prepare data first and then use the template for a whole HTML instead of doing it in parts?